design/loadbalancer-hash-policy-design.md

```diff
@@ -0,0 +1,185 @@
+A client may send a consistent identifier in a header value in order to specify the same upstream service instance should handle all requests with that specific value.
+In addition, [supporting multiple hash policies](https://github.com/projectcontour/contour/issues/3044) can give even more flexibility.
+If a particular load balancing attribute is not available on a request, users can specify fallback policies on additional attributes to still segment traffic between backends, rather than having Envoy continue on to route requests using the round-robin or random mechanisms.
+
+## Goals
+- Contour should offer the ability to load balance requests based on HTTP user specified request headers
+- The design should offer the ability to configure multiple hash policies (of which specific new types may be implemented separately from this design)
+
+## Non Goals
+- Provide advanced load balancing configuration for `ExtensionService` or `TCPRoute`
+- Provide users direct access to Envoy configuration fields/structures
+
+## High-Level Design
+In order to support multiple hash policies on different attributes of a request, we will add support for a new load balancing strategy on the `LoadBalancerPolicy` object that allows users creating `Route` entries on `HTTPProxy` resources to specify a list of hash policies.
+This new strategy will only be usable for `LoadBalancerPolicy` objects on `HTTPProxy` resources and will be replaced with the default strategy when selected on `TCPProxy` and `ExtensionService` resources.
+Initially, we will only support the option to configure hashing of HTTP request headers selected by name but this may be expanded on in the future as user requests come in.
+
+## Detailed Design
+
+### Changes to `LoadBalancerPolicy`
+The `Strategy` field of the `LoadBalancerPolicy` object will support a new value, `RequestHash` which will denote that request attributes will be hashed by Envoy to make a decision about an upstream cluster instance to route a request to.
+If the `RequestHash` strategy is chosen, Contour will inspect the new `RequestHashPolicies` list field of the `LoadBalancerPolicy` object to build the Envoy hash policy configuration.
+
+```
+type LoadBalancerPolicy struct {
+    // Now also supports `RequestHash`
+	Strategy string `json:"strategy,omitempty"`
+
+	RequestHashPolicies []RequestHashPolicy `json:"requestHashPolicies,omitempty"`
+}
+```
+
+This field will be a list of `RequestHashPolicy` objects with each element holding configuration for an individual hash policy.
+Each list element will allow users configuring a `Route` on a `HTTPProxy` to specify a separate request attribute for Envoy to hash in order to make an upstream load balancing decision.
+`RequestHashPolicy` contains a field `RequestAttribute` specifying which type of request attribute it is targeting.
+If `RequestAttribute` is empty or an unknown value, this hash policy entry will be ignored.
```

Comment:
what are the valid values? Just `Header` for now?

Reply:
just `Header` for now, could be expanded to `Cookie` etc. from here: https://www.envoyproxy.io/docs/envoy/v1.17.0/api-v3/config/route/v3/route_components.proto#config-route-v3-routeaction-hashpolicy